src/ui/profile/ProfileSelection.js

```diff
@@ -0,0 +1,170 @@
+// eslint-disable-next-line no-redeclare
+/* global chrome */
+import React, { useEffect, useState } from 'react';
+import PropTypes from 'prop-types';
+import styles from '../styles/profile.module.css';
+import AddProfile from './AddProfile';
+import BeyondLoader from './BeyondLoader';
+import Profiles from './Profile';
+
+const LoadedProfile = (profileModel) => ({
+  status: 'loaded',
+  profile: profileModel,
+});
+
+const ErrorProfile = (error, profile) => ({
+  status: 'error',
+  error,
+  profile,
+});
+
+const LoadingProfile = (profile) => ({
+  status: 'loading',
+  profile,
+});
+
+function ProfileSelection(props) {
+  const { onCharacterReady } = props;
+  const [profiles, setProfiles] = useState(new Map());
+  const [currentProfile, setCurrentProfile] = useState(null);
+  const [errorMessage, setErrorMessage] = useState('');
+
+  useEffect(() => {
+    chrome.storage.local.get('savedProfiles', (result) => {
+      if (!result.savedProfiles) return;
+
+      const keyPairs = result.savedProfiles.map((profile) => [profile.id, LoadedProfile(profile)]);
+      setProfiles(new Map(keyPairs));
+    });
+  }, []);
+
+  const onAddProfile = (id, profilePromise) => {
+    if (profiles.has(id)) {
+      throw new Error('ALREADY_EXISTS');
+    }
+
+    // Add placeholder profile with id until loaded
+    setProfiles((map) => {
+      const newMap = new Map(Array.from(map.entries()));
```

Comment:
This was removed in my suggestion above, but there is a simpler way to copy a map:

```js
const newMap = new Map([...map]);
```

For more explanation:

This uses array spread syntax, which you've used before, but here it's doing more than just copying an array.

The reason that you needed to call `Array.from()` on `map.entries()` is that `map.entries()` doesn't return an array, it returns an `Iterator`. These work effectively the same as Rust iterators. `Array.from()` takes any "array-like" object and converts it to an array. For iterators, it simply enumerates the entire iterator into an array and returns it.

JS also has another type called an `Iterable` (effectively equivalent to an `Into<Iterator>` in Rust), which is just an object that can create an `Iterator`. There is a specific method that an object must have in order to be an `Iterable`.

There are a few syntactic constructs in JS that make use of `Iterable`s: the `for-of` loop is only usable on `Iterable`s, and array spread syntax is only usable on `Iterable`s. It turns out that a `Map` is an `Iterable`. It returns an `Iterator` that yields its entries (`[key, value]`). So `new Map([...map])` will create the entry `Iterator` and enumerate all of the entries into a new array, using that to create a new `Map`.